test/ecto/adapters/mssql_test.exs

```diff
@@ -0,0 +1,1240 @@
+defmodule Ecto.Adapters.MsSqlTest do
+  use ExUnit.Case, async: true
+
+  import Ecto.Query
+
+  alias Ecto.Queryable
+  alias Ecto.Adapters.MsSql.Connection, as: SQL
+  alias Ecto.Migration.Reference
+
+  defmodule Model do
+    use Ecto.Schema
+
+    schema "model" do
+      field :x, :integer
+      field :y, :integer
+      field :z, :integer
+      field :w, :decimal
+
+      has_many :comments, Ecto.Adapters.MsSqlTest.Model2,
+        references: :x,
+        foreign_key: :z
+
+      has_one :permalink, Ecto.Adapters.MsSqlTest.Model3,
+        references: :y,
+        foreign_key: :id
+    end
+  end
+
+  defmodule Model2 do
+    use Ecto.Schema
+
+    import Ecto.Query
+
+    schema "model2" do
+      belongs_to :post, Ecto.Adapters.MsSqlTest.Model,
+        references: :x,
+        foreign_key: :z
+    end
+  end
+
+  defmodule Model3 do
+    use Ecto.Schema
+
+    import Ecto.Query
+
+    @schema_prefix "foo"
+    schema "model3" do
+      field :binary, :binary
+    end
+  end
+
+  defp plan(query, operation \\ :all) do
+    {query, _} = Ecto.Adapter.Queryable.plan_query(operation, Ecto.Adapters.MsSql, query)
+    query
+  end
+
+  defp all(query), do: query |> SQL.all() |> IO.iodata_to_binary()
+  defp update_all(query), do: query |> SQL.update_all() |> IO.iodata_to_binary()
+  defp delete_all(query), do: query |> SQL.delete_all() |> IO.iodata_to_binary()
+  defp execute_ddl(query), do: query |> SQL.execute_ddl() |> Enum.map(&IO.iodata_to_binary/1)
+
+  defp insert(prefx, table, header, rows, on_conflict, returning) do
+    IO.iodata_to_binary(SQL.insert(prefx, table, header, rows, on_conflict, returning))
+  end
+
+  defp update(prefx, table, fields, filter, returning) do
+    IO.iodata_to_binary(SQL.update(prefx, table, fields, filter, returning))
+  end
+
+  defp delete(prefx, table, filter, returning) do
+    IO.iodata_to_binary(SQL.delete(prefx, table, filter, returning))
+  end
+
+  test "from" do
```

Comment:
Btw, have you recently given a full pass comparing this file to "pgsql_test.exs" to see if there were any additions? If not, it is not problem, I can do that after merge. It would actually be a great opportunity for me to learn more about MSSQL.

Reply:
I did, and I presume this commit is the [reason](https://github.com/elixir-ecto/ecto_sql/commit/d3a3f1b81c723276b972efe73aee865845637db5) but I haven't checked why it is failing

```
  1) test update all with casting and dumping (Ecto.Integration.RepoTest)
     deps/ecto/integration_test/cases/repo.exs:999
     ** (Postgrex.Error) ERROR 26000 (invalid_sql_statement_name) prepared statement "ecto_insert_posts" does not exist
     code: assert %Post{id: id} = TestRepo.insert!(%Post{})
     stacktrace:
       (ecto_sql) lib/ecto/adapters/sql.ex:612: Ecto.Adapters.SQL.raise_sql_call_error/1
       (ecto) lib/ecto/repo/schema.ex:655: Ecto.Repo.Schema.apply/4
       (ecto) lib/ecto/repo/schema.ex:263: anonymous fn/15 in Ecto.Repo.Schema.do_insert/4
       (ecto) lib/ecto/repo/schema.ex:164: Ecto.Repo.Schema.insert!/4
       deps/ecto/integration_test/cases/repo.exs:1002: (test)
```

Reply:
Ah, that failure has been happening on CI for a while. It is unrelated but I cannot reproduce it locally.

Reply:
@josevalim Still no luck with that problem? That's really weird … On my machine I can reproduce it really easily and recently I have confirmed it even testing `Ubuntu Server` in `VirtualBox` … Do you think that `.vdi` file would be helpful?

Reply:
I'm not using VirtualBox but here is my two cents, this is probably issue with caching in postgrex, and it depends in which order tests are run since I'm getting random results. Sometimes it is just logged single error 

`11:06:31.321 [error] Postgrex.Protocol (#PID<0.314.0>) disconnected: ** (DBConnection.ConnectionError) connection was checked out with status :idle but it was checked in with status :transaction`

sometimes it is mentioned exception above.

Is it possible that one test deleted cache while other passed prepare (cash found) and next time it called actual query is not found in the cache?

Reply:
I got this info in logs too (sometimes)

```
11:06:04.166 [info]  Postgrex.Protocol (#PID<0.317.0>) disconnected: ** (DBConnection.ConnectionError) client #PID<0.1486.0> exited
```